src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,577 @@
+import jdk.internal.vm.annotation.ForceInline;
+
+import java.util.*;
+import java.util.function.BiConsumer;
+import java.util.function.BinaryOperator;
+import java.util.function.Supplier;
+
+
+/**
+ * An intermediate operation that transforms a stream of input elements into a
+ * stream of output elements, optionally applying a final action when the end of
+ * the upstream is reached. The transformation may be stateless or stateful,
+ * and a Gatherer may buffer arbitrarily much input before producing any output.
+ *
+ * <p>Gatherer operations can be performed either sequentially,
+ * or be parallelized -- if a combiner function is supplied.
+ *
+ * <p>Examples of gathering operations include, but is not limited to:
+ * grouping elements into batches (windowing functions);
+ * de-duplicating consecutively similar elements; incremental accumulation
+ * functions (prefix scan); incremental reordering functions, etc.  The class
+ * {@link java.util.stream.Gatherers} provides implementations of common
+ * gathering operations.
+ *
+ * @apiNote
+ * <p>A {@code Gatherer} is specified by four functions that work together to
+ * process input elements, optionally using intermediate state, and optionally
+ * perform a final action at the end of input.  They are: <ul>
+ *     <li>creating a new, potentially mutable, state ({@link #initializer()}
+ *     )</li>
+ *     <li>integrating a new input element ({@link #integrator()})</li>
+ *     <li>combining two states into one ({@link #combiner()})</li>
+ *     <li>performing an optional final action ({@link #finisher()})</li>
+ * </ul>
+ *
+ * <p>Each invocation to {@link #initializer()}, {@link #integrator()},
+ * {@link #combiner()}, and {@link #finisher()} must return a semantically
+ * identical result.
+ *
+ * <p>Implementations of Gatherer must not capture, retain, or expose to
+ * other threads, the references to the state instance, or the downstream
+ * {@link Downstream} for longer than the invocation duration of the method
+ * which they are passed to.
+ *
+ * <p>Performing a gathering operation with a {@code Gatherer} should produce a
+ * result equivalent to:
+ *
+ * {@snippet lang = java:
+ * Gatherer.Downstream<? super R> downstream = ...;
+ * A state = gatherer.initializer().get();
+ * for (T t : data) {
+ *     gatherer.integrator().integrate(state, t, downstream);
+ * }
+ * gatherer.finisher().accept(state, downstream);
+ * }
+ *
+ * <p>However, the library is free to partition the input, perform the
+ * integrations on the partitions, and then use the combiner function to
+ * combine the partial results to achieve a gathering operation.  (Depending
+ * on the specific gathering operation, this may perform better or worse,
+ * depending on the relative cost of the integrator and combiner functions.)
+ *
+ * <p>In addition to the predefined implementations in {@link Gatherers}, the
+ * static factory methods {@code of(...)} and {@code ofSequential(...)}
```

Comment:
I assume both of these could be links.

Reply:
Unfortunately they can't, since they are both overloaded and here we're talking about them "generically".